tools/CopyrightDateTool.scala

```diff
@@ -0,0 +1,174 @@
+      fileLastUpdateYear.toInt != gitLastUpdateYear.toInt
+    }
+
+    def ifUpdateDateWrongStr(): Boolean = {
+      fileLastUpdateYear != gitLastUpdateYear
+    }
+
+    def ifOneYear(): Boolean = {
+      gitCreationYear != gitLastUpdateYear
+    }
+  }
+
+  case class FileDates(pathToFile: String,
+                      fileLastUpdateYear: String,
+                      gitLastUpdateYear: String,
+                      fileCreationYear: String,
+                      gitCreationYear: String) extends DateComparison
+
+  case class GitInfo(message: String, date: String)
+
+  def readFilesList(): Array[String] = {
+    val myval =  (s"find . -name *.scala ! -path ./tools/*  -type f") !!
+    val files_array = myval.split("\n").map(_.toString).distinct
+    files_array
+  }
+
+  def readFile(pathToFile: String): String = {
+    val source = scala.io.Source.fromFile(pathToFile)("UTF-8")
+    val lines = try source.mkString finally source.close()
+    lines
+  }
+
+  def ifFileWasInGit(pathToFile: String): Boolean = {
+    val res_dates = (s"git log $pathToFile" !!)
+    res_dates.nonEmpty
+  }
+
+  def getGitCreationYear(pathToFile: String): String = {
+    val buf_dates = (s"git log   --format=%ci $pathToFile" !!)
+    val dates = buf_dates.split("\\n").distinct
+    val res = dates.last.split("-")(0)
+    res
+  }
+
+  def getGitLastUpdateYear(pathToFile: String): String = {
+    val buf_dates = (s"git log -3  --format=%cd $pathToFile" !!)
+    val buf_messages = (s"git log -3  $pathToFile" !!)
+    val messages = buf_messages.split("commit").distinct.drop(1)
+    val dates = buf_dates.split("\\n").distinct
+    val gits_info = (messages, dates).zipped.map{ case(a,b) => GitInfo(a,b)}
+    val res = gits_info.filter(!_.message.contains("[Sync copyright dates]"))
+    val fifth_column = 4
+    val first = 0
+    res(first).date.split(" ")(fifth_column)
+  }
+
+  def getYear(pathToFile: String, lower_bound: Int, upper_bound: Int): String = {
+    val file = readFile(pathToFile)
+    val third_line = 2
+    val start_index = file.split("\\n")(third_line).indexOf("(c) 20")
+    val year = file.split("\\n")(third_line)
+                   .substring(start_index + lower_bound,
+                              start_index + upper_bound)
+    year
+  }
+
+  def getFileCreationYear(pathToFile: String): String = {
+    val lb = "(c) ".length()
+    val ub = "(c) 20XX".length()
+    val fileCreationYear = getYear(pathToFile, lb, ub)
+    fileCreationYear
+  }
+
+  def getFileLastUpdateYear(pathToFile: String): String = {
+    val lb = "(c) 20XX-".length()
+    val ub = "(c) 20XX-XXXX".length()
+    val fileLastUpdateYear = getYear(pathToFile, lb, ub)
+    fileLastUpdateYear
+  }
+
+  def analyzeDates(): Array[FileDates] = {
+    val files_array = readFilesList()
+    var data = Array[FileDates]()
+    for(file_path <- files_array) {
+      if(readFile(file_path).contains("|    (c) 20") && ifFileWasInGit(file_path)) {
```

Comment:
This could be simplified by starting with a list of Scala files that are in the repo instead of starting with a list of Scala files that are in the working directory and then skipping the files that are not in the repo.

This gives a list of *.scala files that are known to be in the repo,

` git ls-files|grep '\.scala$'`